File: src/selmer/node.py
from __future__ import annotations

from collections.abc import Callable
from dataclasses import dataclass
from typing import Any, Protocol

from .types import Context


class FunctionHandler(Protocol):
    meta: dict[str, Any]

    def __call__(self, context: Context) -> Any: ...


@dataclass
class FunctionNode:
    handler: Callable[[Context], Any]

    def render_node(self, context: Context) -> Any:
        return self.handler(context)

    @property
    def meta(self) -> dict[str, Any] | None:
        return getattr(self.handler, "meta", None)


@dataclass
class TextNode:
    text: str

    def render_node(self, context: Context) -> str:
        return str(self.text)

    def __str__(self) -> str:
        return str(self.text)

File: src/node.ts
import type { Context, FunctionHandler, TemplateNode } from "./types.js";

export class FunctionNode implements TemplateNode {
  public readonly handler: FunctionHandler;

  constructor(handler: FunctionHandler) {
    this.handler = handler;
  }

  renderNode(context: Context): unknown {
    return this.handler(context);
  }

  get meta(): FunctionHandler["meta"] {
    return this.handler.meta;
  }
}

export class TextNode implements TemplateNode {
  public readonly text: string;

  constructor(text: string) {
    this.text = text;
  }

  renderNode(_context: Context): string {
    return String(this.text);
  }

  toString(): string {
    return String(this.text);
  }
}
